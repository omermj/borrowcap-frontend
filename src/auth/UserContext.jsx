import React from "react";

/** Context to provide currentUser thoughout the app */

const UserContext = React.createContext();

export default UserContext;
