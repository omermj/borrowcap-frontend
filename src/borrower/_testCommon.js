import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = () => {
  // Mock api call with active requests
  BorrowcapApi.getLoanRequestsByBorrowerId.mockResolvedValue([
    {
      id: 4,
      amtRequested: 10000,
      purpose: "Business",
      appOpenDate: "2021/01/01",
      interestRate: 0.05,
      term: 12,
      installmentAmt: 1000,
    },
    {
      id: 5,
      amtRequested: 20000,
      purpose: "Home",
      appOpenDate: "2021/02/01",
      interestRate: 0.07,
      term: 24,
      installmentAmt: 2000,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
