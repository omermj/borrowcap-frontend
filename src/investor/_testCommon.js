import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getFundedLoansByUserId.mockResolvedValue({
    investor: [
      {
        id: 1,
        investorId: 3,
        amtInvested: "5000",
        amtFunded: "5000",
        fundedDate: "2023/01/01",
        interestRate: "0.05",
        term: 12,
        installmentAmt: "428.04",
        remainingBalance: "5000",
      },
    ],
  });
  await BorrowcapApi.getAvailableInvestments.mockResolvedValue([
    {
      id: 7,
      amtApproved: "8000",
      amtFunded: "0",
      purpose: "Medical",
      approvedDate: "2023/01/01",
      fundingDeadline: "2023/02/1",
      interestRate: "0.07",
      term: 24,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
