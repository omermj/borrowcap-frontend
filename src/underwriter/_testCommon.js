import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getLoanRequests.mockResolvedValue([
    {
      id: 2,
      borrowerId: 2,
      amtRequested: "10000",
      purpose: "Education",
      purposeId: 3,
      appOpenDate: "2023/01/01",
      interestRate: "0.07",
      term: 36,
      installmentAmt: "319.86",
    },
  ]);

  await BorrowcapApi.getApprovedRequests.mockResolvedValue([
    {
      id: 1,
      borrowerId: 2,
      amtRequested: "10000",
      amtApproved: "9000",
      amtFunded: "5000",
      purpose: "Car",
      purposeId: 2,
      appOpenDate: "2023/01/01",
      appApprovedDate: "2023/01/05",
      fundingDeadline: "2023/02/05",
      interestRate: "0.05",
      term: 12,
      installmentAmt: "428.04",
      availableForFunding: false,
      isFunded: true,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
