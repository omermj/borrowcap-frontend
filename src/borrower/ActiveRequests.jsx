import { useState, useEffect, useContext } from "react";
import TableComponent from "../common/TableComponent";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import BorrowcapApi from "../api/api";
import UserContext from "../auth/UserContext";

/** Table showing all Active Loan Requests for logged in user */

const ActiveRequests = () => {
  const { currentUser } = useContext(UserContext);
  const [loanRequests, setLoanRequests] = useState([]);

  // Define table headers with labels and formatters
  const headers = {
    id: { label: "ID", formatter: "none" },
    amtRequested: {
      label: "Requested Amount",
      formatter: formatCurrency,
    },
    purpose: { label: "Purpose", formatter: "none" },
    appOpenDate: { label: "Application Date", formatter: formatDate },
    interestRate: { label: "Interest Rate", formatter: formatPercent },
    term: { label: "Term", formatter: "none" },
    installmentAmt: { label: "Installment", formatter: formatCurrency },
  };

  // get active loan requests on initial render
  useEffect(() => {
    async function fetchLoanRequests() {
      const requests = await BorrowcapApi.getLoanRequestsByBorrowerId(
        currentUser.id
      );
      setLoanRequests(requests);
    }
    fetchLoanRequests();
  }, [currentUser.id]);

  if (!loanRequests.length) return <div>No active loan requests.</div>;

  return (
    <div className="border mb-4">
      <div className="py-2">
        <span className="align-middle h5">New Loan Requests</span>
      </div>
      <TableComponent headers={headers} tableData={loanRequests} />
    </div>
  );
};

export default ActiveRequests;
