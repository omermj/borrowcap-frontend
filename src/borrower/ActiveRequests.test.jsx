import { render, waitFor, screen } from "@testing-library/react";
import ActiveRequests from "./ActiveRequests";
import { expect, it } from "vitest";
import UserContext from "../auth/UserContext";
import { commonBeforeEach, testUser } from "./_testCommon.js";
import BorrowcapApi from "../api/api.js";

// Mock api calls
vi.mock("../api/api");
beforeEach(commonBeforeEach);

// Smoke test
it("renders without crashing", () => {
  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <ActiveRequests />
    </UserContext.Provider>
  );
});

// Test table rendering
it("renders table with loan requests", async () => {
  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <ActiveRequests />
    </UserContext.Provider>
  );

  // Wait for api call to resolve and ppulate table
  await waitFor(() => {
    expect(BorrowcapApi.getLoanRequestsByBorrowerId).toHaveBeenCalledWith(
      testUser.id
    );
  });

  // Verify that the table is populated
  expect(screen.getByText("New Loan Requests")).toBeInTheDocument();
  expect(screen.getByText("Requested Amount")).toBeInTheDocument();
  expect(screen.getByText("Purpose")).toBeInTheDocument();
  expect(screen.getByText("Application Date")).toBeInTheDocument();
  expect(screen.getByText("Interest Rate")).toBeInTheDocument();
  expect(screen.getByText("Term")).toBeInTheDocument();
  expect(screen.getByText("Installment")).toBeInTheDocument();

  // Verify that the table has the correct data
  expect(screen.getByText("$10,000.00")).toBeInTheDocument();
  expect(screen.getByText("Business")).toBeInTheDocument();
  expect(screen.getByText("2021-01-01")).toBeInTheDocument();
  expect(screen.getByText("5.00%")).toBeInTheDocument();
  expect(screen.getByText("12")).toBeInTheDocument();
  expect(screen.getByText("$1,000.00")).toBeInTheDocument();
});

// Test table rendering with no loan requests
it("renders table with no loan requests", async () => {
  // Mock api call to return empty array
  BorrowcapApi.getLoanRequestsByBorrowerId.mockResolvedValue([]);

  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <ActiveRequests />
    </UserContext.Provider>
  );

  // Wait for api call to resolve and ppulate table
  await waitFor(() => {
    expect(BorrowcapApi.getLoanRequestsByBorrowerId).toHaveBeenCalledWith(
      testUser.id
    );
  });

  expect(screen.getByText("No active loan requests.")).toBeInTheDocument();
});

// Snapshot test
it("matches snapshot", () => {
  const { asFragment } = render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <ActiveRequests />
    </UserContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
