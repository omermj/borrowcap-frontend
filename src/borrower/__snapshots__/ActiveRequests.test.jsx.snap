// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    No active loan requests.
  </div>
</DocumentFragment>
`;
